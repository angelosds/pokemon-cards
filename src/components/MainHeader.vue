<template>
  <header class="main-header">
    <div class="main-header__content">
      <span class="main-header__title">
        <slot />
      </span>

      <button class="main-header__button" type="button" @click="onSearchClick">
        <svg
          viewBox="0 0 24 24"
          fill="none"
          stroke="currentColor"
          stroke-width="2"
          stroke-linecap="round"
          stroke-linejoin="round"
          class="main-header__icon"
        >
          <circle cx="11" cy="11" r="8"></circle>
          <line x1="21" y1="21" x2="16.65" y2="16.65"></line>
        </svg>
      </button>
    </div>

    <div class="main-header__search" v-if="isSearchActive">
      <button class="main-header__close" type="button" @click="onCloseClick">
        &times;
      </button>

      <form @submit.prevent="onSubmit">
        <input
          type="search"
          class="main-header__input"
          :placeholder="placeholder"
          v-model="searchText"
        />
      </form>
    </div>
  </header>
</template>

<script lang="ts">
import { Component, Prop, Vue } from "vue-property-decorator";

@Component
export default class MainHeader extends Vue {
  @Prop() private readonly placeholder?: string;

  private isSearchActive = false;
  private searchText = "";

  private onCloseClick() {
    this.isSearchActive = false;
  }

  private onSearchClick() {
    this.isSearchActive = true;
  }

  private onSubmit() {
    const { searchText } = this;

    this.isSearchActive = !searchText.trim();

    if (searchText.trim()) {
      this.searchText = "";
      this.$emit("on-search", searchText);
    }
  }
}
</script>

<style lang="scss" scoped>
.main-header {
  background-color: $header-background-color;
  height: 96px;
  padding: 32px 16px;

  &__button,
  &__close {
    align-items: center;
    background-color: transparent;
    border-radius: 8px;
    border: 0;
    color: $primary-color;
    display: flex;
    height: 40px;
    justify-content: center;
    outline: 0;
    padding: 0;
    width: 40px;

    &:hover {
      background-color: $primary-color;
      color: $text-color-inverse;
    }
  }

  &__close {
    font-size: 32px;
    justify-self: end;
  }

  &__content {
    align-items: center;
    display: grid;
    grid-gap: 16px;
    grid-template-columns: 1fr 40px;
    justify-content: center;
  }

  &__icon {
    height: 24px;
    width: 24px;
  }

  &__input {
    background-color: transparent;
    border-bottom: 4px solid $primary-color;
    border-left: 0;
    border-right: 0;
    border-top: 0;
    height: 40px;
    outline: 0;
    padding: 4px;
    width: 100%;
  }

  &__title {
    font-size: 2.2rem;
    font-weight: 700;
    justify-self: center;
  }

  &__search {
    background-color: $search-background-color;
    display: grid;
    grid-gap: 16px;
    grid-template-rows: 40px 1fr;
    height: 100vh;
    left: 0;
    padding: 32px 16px;
    position: fixed;
    top: 0;
    width: 100vw;
  }
}
</style>
