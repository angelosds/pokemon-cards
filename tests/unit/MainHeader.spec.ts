import { shallowMount } from "@vue/test-utils";

import MainHeader from "@/components/MainHeader.vue";

describe("MainHeader.vue", () => {
  const build = (propsData?: object, slots?: object) => {
    const wrapper = shallowMount(MainHeader, {
      slots: { ...slots },
      propsData: { ...propsData },
      stubs: ["router-link"]
    });

    const closeButton = () => wrapper.find(".main-header__close");
    const searchButton = () => wrapper.find(".main-header__button");
    const searchInput = () => wrapper.find(".main-header__input");
    const title = () => wrapper.find(".main-header__title");

    return { closeButton, searchButton, searchInput, title, wrapper };
  };

  it("renders the component", () => {
    const { wrapper } = build();

    expect(wrapper).toMatchSnapshot();
  });

  it("renders a title in a default slot", () => {
    const titleText = "Pokémon cards";
    const { title } = build(undefined, { default: titleText });

    expect(title().text()).toMatch(titleText);
  });

  it("does not shows search input by default", () => {
    const { searchInput } = build();

    expect(searchInput().exists()).toBe(false);
  });

  it("shows search input when click on search button", async () => {
    const { searchButton, searchInput } = build();

    expect(searchInput().exists()).toBe(false);

    await searchButton().trigger("click");

    expect(searchInput().exists()).toBe(true);
  });

  it("hides search input when click on close button", async () => {
    const { closeButton, searchButton, searchInput } = build();

    expect(searchInput().exists()).toBe(false);

    await searchButton().trigger("click");

    expect(searchInput().exists()).toBe(true);

    await closeButton().trigger("click");

    expect(searchInput().exists()).toBe(false);
  });

  it("renders a input placeholder", async () => {
    const placeholderText = "Search by pokémon name";
    const { searchButton, searchInput } = build({
      placeholder: placeholderText
    });

    await searchButton().trigger("click");

    expect(searchInput().attributes("placeholder")).toMatch(placeholderText);
  });

  it("does not emits on-search when submit an empty form", async () => {
    const pokemonName = "";
    const { searchButton, searchInput, wrapper } = build();

    await searchButton().trigger("click");
    await searchInput().setValue(pokemonName);
    await wrapper.find("form").trigger("submit");

    expect(wrapper.emitted("on-search")).toBeFalsy();
  });

  it("emits on-search when submit the form", async () => {
    const pokemonName = "pikachu";
    const { searchButton, searchInput, wrapper } = build();

    await searchButton().trigger("click");
    await searchInput().setValue(pokemonName);
    await wrapper.find("form").trigger("submit");

    expect(wrapper.emitted("on-search")).toBeTruthy();
    expect(wrapper.emitted("on-search")?.[0]?.[0]).toEqual(pokemonName);
  });
});
