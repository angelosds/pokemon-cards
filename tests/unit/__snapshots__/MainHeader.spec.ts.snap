// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainHeader.vue renders the component 1`] = `
<header class="main-header">
  <div class="main-header__content">
    <router-link to="[object Object]" class="main-header__title"></router-link> <button type="button" class="main-header__button"><svg viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="main-header__icon">
        <circle cx="11" cy="11" r="8"></circle>
        <line x1="21" y1="21" x2="16.65" y2="16.65"></line>
      </svg></button>
  </div>
  <!---->
</header>
`;
